lang/cs/Org.Apache.REEF.Network/Group/Topology/TreeTopology.cs

```diff
@@ -0,0 +1,215 @@
+                    parentId);
+
+            //add all its children
+            foreach (TaskNode childNode in selfTaskNode.GetChildren())
+            {
+                confBuilder.BindSetEntry<MpiConfigurationOptions.TopologyChildTaskIds, string>(
+                    GenericType<MpiConfigurationOptions.TopologyChildTaskIds>.Class,
+                    childNode.TaskId);
+            }
+
+            if (OperatorSpec is BroadcastOperatorSpec<T>)
+            {
+                BroadcastOperatorSpec<T> broadcastSpec = OperatorSpec as BroadcastOperatorSpec<T>;
+                if (taskId.Equals(broadcastSpec.SenderId))
+                {
+                    confBuilder.BindImplementation(GenericType<IMpiOperator<T>>.Class, GenericType<BroadcastSender<T>>.Class);
+                }
+                else
+                {
+                    confBuilder.BindImplementation(GenericType<IMpiOperator<T>>.Class, GenericType<BroadcastReceiver<T>>.Class);
+                }
+            }
+            else if (OperatorSpec is ReduceOperatorSpec<T>)
+            {
+                ReduceOperatorSpec<T> reduceSpec = OperatorSpec as ReduceOperatorSpec<T>;
+                confBuilder.BindImplementation(typeof(IReduceFunction<T>), reduceSpec.ReduceFunction.GetType());
+
+                if (taskId.Equals(reduceSpec.ReceiverId))
+                {
+                    confBuilder.BindImplementation(GenericType<IMpiOperator<T>>.Class, GenericType<ReduceReceiver<T>>.Class);
+                }
+                else
+                {
+                    confBuilder.BindImplementation(GenericType<IMpiOperator<T>>.Class, GenericType<ReduceSender<T>>.Class);
+                }
+            }
+            else if (OperatorSpec is ScatterOperatorSpec<T>)
+            {
+                ScatterOperatorSpec<T> scatterSpec = OperatorSpec as ScatterOperatorSpec<T>;
+                if (taskId.Equals(scatterSpec.SenderId))
+                {
+                    confBuilder.BindImplementation(GenericType<IMpiOperator<T>>.Class, GenericType<ScatterSender<T>>.Class);
+                }
+                else
+                {
+                    confBuilder.BindImplementation(GenericType<IMpiOperator<T>>.Class, GenericType<ScatterReceiver<T>>.Class);
+                }
+            }
+            else
+            {
+                throw new NotSupportedException("Spec type not supported");
+            }
+
+            return confBuilder.Build();
+        }
+
+        public void AddTask(string taskId)
+        {
+            if (string.IsNullOrEmpty(taskId))
+            {
+                throw new ArgumentNullException("taskId");
+            }
+            else if (_nodes.ContainsKey(taskId))
+            {
+                throw new ArgumentException("Task has already been added to the topology");
+            }
+
+            if (taskId.Equals(_rootId))
+            {
+                SetRootNode(_rootId);
```

Comment:
I think it probably make more sense to just SetRooNode in the TreeTopology ctor, and just report an error here.